// @captify/admin/src/services/debug.ts
/**
 * Debug service for clearing caches and testing
 */

export async function execute(
  request: {
    service: string;
    operation: string;
    data?: any;
  },
  credentials?: any
) {
  try {
    const { operation } = request;

    switch (operation) {
      case "clearCredentialsCache": {
        // This will call the server-side cache clearing
        return {
          success: true,
          data: { message: "Cache clear request sent" },
          metadata: {
            requestId: `debug-clear-cache-${Date.now()}`,
            timestamp: new Date().toISOString(),
            source: "debug.clearCredentialsCache",
          },
        };
      }

      default:
        return {
          success: false,
          error: `Unsupported debug operation: ${operation}`,
          metadata: {
            requestId: `debug-error-${Date.now()}`,
            timestamp: new Date().toISOString(),
            source: "debug.execute",
          },
        };
    }
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : "Debug operation failed",
      metadata: {
        requestId: `debug-error-${Date.now()}`,
        timestamp: new Date().toISOString(),
        source: "debug.execute",
      },
    };
  }
}
